> **Credits:** Many thanks to **Liran Nahum** for his contributions to this chapter.

In [66]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

%matplotlib inline

[_Pandas_][1] is the most exahustive library for data analysis in Python, and it can be described as the Pythonic son of SQL and Excel. It is based on [*NumPy*](https://www.numpy.org/) and [*matplotlib*](https://matplotlib.org/), but provides a unified API for many data-related operations, including manipulation, I/O, indexing, alignment, vizualizations, pivoting, grouping, etc.. In addition, it provides an impressive set of tools for working with time-series.

[1]: http://pandas.pydata.org/ "pandas org"

> **Note:** NumPy is the mathematical infrastructure of pandas, and although most of our work will be done with pandas, it is a good idea to get acquainted with numpy basics through the [chapter about numpy](https://drive.google.com/drive/folders/1YzgjBy8ETzWWHrnYThl9qA_qRH8AcA-o?usp=sharing). More specifically, one should be familiar with the concepts of numpy array, advanced indexing and broadcasting.

# The _Series_ class

## Basics

_Series_ objects should be thought of as table columns. Although they are not exactly like _NumPy's_ 1-dimensional arrays, they are very similar, and behave similarly in many cases. _Series_ objects have many attributes and support many methods, which are all documented alphabetically [here][1] and by subject [here][2].

The elements of the series (and the rows of the data-frame, as we'll see later) are **labeled** by an important attribute called **_index_**, which is an instance of the class [`Index`][3] (or one of its subclasses). If _index_ is not specified, then the elements are labeled by the regular row-counter, and the _index_ of the _series_ contains the integers from 0 to _n-1_.

[1]: http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.html "Series API documentation A-Z"
[2]: http://pandas.pydata.org/pandas-docs/stable/api.html#series "Series API documentation"
[3]: http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Index.html "Index API documentation"

In [67]:
dudes_ages = pd.Series([24, 25, 20, 24, 19])
print(dudes_ages)

0    24
1    25
2    20
3    24
4    19
dtype: int64


In [68]:
dudes_ages

0    24
1    25
2    20
3    24
4    19
dtype: int64

In [69]:
type(dudes_ages)

pandas.core.series.Series

In [70]:
pd.Series([24, 25, 20, 24, 19, 199.35, "123.4"])

0        24
1        25
2        20
3        24
4        19
5    199.35
6     123.4
dtype: object

In [71]:
pd.Series([24, 25, 20, 24, 19, 199.35, "123.4"], dtype="float32")

0     24.000000
1     25.000000
2     20.000000
3     24.000000
4     19.000000
5    199.350006
6    123.400002
dtype: float32

We can also specify the index.

In [72]:
pd.Series?

In [73]:
dudes_ages = pd.Series(
    [24, 25, 20, 24, 19],
    index=['Avi', 'Ben', 'Gad', 'Ben', 'Hen']
)
print(dudes_ages)

Avi    24
Ben    25
Gad    20
Ben    24
Hen    19
dtype: int64


In [74]:
dudes_ages / 2

Avi    12.0
Ben    12.5
Gad    10.0
Ben    12.0
Hen     9.5
dtype: float64

In [75]:
dudes_ages == 20 # we got new series when appling vecor masks

Avi    False
Ben    False
Gad     True
Ben    False
Hen    False
dtype: bool

> **Note:** The index values do **NOT** have to be unique. This is actually a very important concept, which has a lot to do with the speed of pandas. We will see its usage below.

Series support broadcasting much like numpy arrays. Note that the index is kept in the result.

In [76]:
dudes_young = dudes_ages < 21
dudes_young

Avi    False
Ben    False
Gad     True
Ben    False
Hen     True
dtype: bool

In [77]:
dudes_ages = dudes_ages + 1
dudes_ages

Avi    25
Ben    26
Gad    21
Ben    25
Hen    20
dtype: int64

In [78]:
data = pd.Series([5, 15, 20, 41, 58, 80, 87],
                 index=['Avi', 'Ben', 'Gad', 'Yair', 'Hen', "Yuval", "Dan"])

print(data * 2)

Avi       10
Ben       30
Gad       40
Yair      82
Hen      116
Yuval    160
Dan      174
dtype: int64


> Your turn:

1. Create a series named 'ser1' with the following data, and print it: 87, 41, 23, 56, 81, 84, 1, 2, 4.
2. Create a series named 'scores' with the same data but add the following keys: 'Avi', 'Ben', 'Gad', 'Yair', 'Hen', "Yuval", "Dan" , "Bob", "Tzvi".
3. Add the value 20 to all scores and print the result (series 'scores').
4. Print True or False for every value that is greater than 50.

In [79]:
'''
1. Create a series named 'ser1' with the following data, and print it: 87, 41, 23, 56, 81, 84, 1, 2, 4.
2. Create a series named 'scores' with the same data but add the following keys: 'Avi', 'Ben', 'Gad', 'Yair', 'Hen', "Yuval", "Dan" , "Bob", "Tzvi".
3. Add the value 20 to all scores and print the result (series 'scores').
4. Print True or False for every value that is greater than 50.
'''

ser1 = pd.Series([87, 41, 23, 56, 81, 84, 1, 2, 4])

scores = pd.Series([87, 41, 23, 56, 81, 84, 1, 2, 4],
                 index=['Avi', 'Ben', 'Gad', 'Yair', 'Hen', "Yuval", "Dan", 'Bob', 'Tzvi'])
scores_plus_20 = scores + 20

print('The result of adding the value 20 to all scores\n', scores + 20)

print("Which score's values greate than 50?\n", scores > 50)


The result of adding the value 20 to all scores
 Avi      107
Ben       61
Gad       43
Yair      76
Hen      101
Yuval    104
Dan       21
Bob       22
Tzvi      24
dtype: int64
Which score's values greate than 50?
 Avi       True
Ben      False
Gad      False
Yair      True
Hen       True
Yuval     True
Dan      False
Bob      False
Tzvi     False
dtype: bool


### Solution

In [80]:
# 1.
ser1 = pd.Series([87, 41, 23, 56, 81, 84, 1, 2, 4])
ser1

0    87
1    41
2    23
3    56
4    81
5    84
6     1
7     2
8     4
dtype: int64

In [81]:
# 2.
scores = pd.Series(
    [87, 41, 23, 56, 81, 84, 1, 2, 4],
    index=['Avi', 'Ben', 'Gad', 'Yair', 'Hen', "Yuval", "Dan" , "Bob", "Tzvi"]
)
scores

Avi      87
Ben      41
Gad      23
Yair     56
Hen      81
Yuval    84
Dan       1
Bob       2
Tzvi      4
dtype: int64

In [82]:
# 3.
print(scores + 20)

Avi      107
Ben       61
Gad       43
Yair      76
Hen      101
Yuval    104
Dan       21
Bob       22
Tzvi      24
dtype: int64


In [83]:
# 4.
print(scores > 50)

Avi       True
Ben      False
Gad      False
Yair      True
Hen       True
Yuval     True
Dan      False
Bob      False
Tzvi     False
dtype: bool


## Indexing and selecting data

Being both array-like and dictionary-like, _Series_ objects intuitively support the standard item-getters of the form `s[label]`, `s.label` and `s[i]`. However, these may be both impractical or ambiguous in some cases, so to provide maximum clarity, *pandas* supports two methods for indexing and selecting data:

* Label-based - implemented by the item-getter `.loc[]`
* integer-position-based - implemented by the item-getter `.iloc[]`


> **Note:** These methods support simple cases, but may be very confusing in more complex scenarios. In such cases it is a good idea to try the thorough documentation for [indexing and selecting data][1]. This is, however, beyond the scope of our course.

[1]: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-and-selecting-data "indexing and selecting data"

In [84]:
names = ['Alon', 'Ari','Barak', 'Itamar','Yarden','Ofer', 'Liran',
         'Oren', 'Ron','Shachar','Shalom','Tal', 'Yosi', 'Tomer',
         'Nir', 'Yaniv', 'Maor', 'Guy', 'Nir', 'Yosi']

cities = ['Beer Sheba', 'Tel Aviv', 'Haifa', 'Rishon LeZion', 'Beersheba',
          'Herzliya', 'Tal-Aviv', 'Rishon-Lezion', 'Jerusalem', 'Petach-Tikva',
          'Petach Tikva', 'Rishon-LeZion', 'Telaviv', 'Beer-7', 'Tel-Aviv',
          'Jerusalem', 'Haifa', 'Rishon LeZion', 'Beer-Sheba', 'Tel-Aviv']

dudes_cities = pd.Series(data=cities, index=names)
print(dudes_cities)

Alon          Beer Sheba
Ari             Tel Aviv
Barak              Haifa
Itamar     Rishon LeZion
Yarden         Beersheba
Ofer            Herzliya
Liran           Tal-Aviv
Oren       Rishon-Lezion
Ron            Jerusalem
Shachar     Petach-Tikva
Shalom      Petach Tikva
Tal        Rishon-LeZion
Yosi             Telaviv
Tomer             Beer-7
Nir             Tel-Aviv
Yaniv          Jerusalem
Maor               Haifa
Guy        Rishon LeZion
Nir           Beer-Sheba
Yosi            Tel-Aviv
dtype: object


In [85]:
dudes_cities["Tal"]

'Rishon-LeZion'

In [86]:
dudes_cities.Yosi

Yosi     Telaviv
Yosi    Tel-Aviv
dtype: object

In [87]:
dudes_cities["Yosi"]

Yosi     Telaviv
Yosi    Tel-Aviv
dtype: object

In [88]:
dudes_cities2 = pd.Series(data=cities).sort_values(ascending=False)
print(dudes_cities2)

12          Telaviv
19         Tel-Aviv
14         Tel-Aviv
1          Tel Aviv
6          Tal-Aviv
7     Rishon-Lezion
11    Rishon-LeZion
17    Rishon LeZion
3     Rishon LeZion
9      Petach-Tikva
10     Petach Tikva
8         Jerusalem
15        Jerusalem
5          Herzliya
16            Haifa
2             Haifa
4         Beersheba
18       Beer-Sheba
13           Beer-7
0        Beer Sheba
dtype: object


In [89]:
dudes_cities2[0]

'Beer Sheba'

In [90]:
dudes_cities2.loc[0]

'Beer Sheba'

In [91]:
dudes_cities2.iloc[0]

'Telaviv'

### _loc_ - Label-based indexing

see more at [Selection by Label][1]

[1]: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#indexing-label

In [92]:
dudes_cities

Alon          Beer Sheba
Ari             Tel Aviv
Barak              Haifa
Itamar     Rishon LeZion
Yarden         Beersheba
Ofer            Herzliya
Liran           Tal-Aviv
Oren       Rishon-Lezion
Ron            Jerusalem
Shachar     Petach-Tikva
Shalom      Petach Tikva
Tal        Rishon-LeZion
Yosi             Telaviv
Tomer             Beer-7
Nir             Tel-Aviv
Yaniv          Jerusalem
Maor               Haifa
Guy        Rishon LeZion
Nir           Beer-Sheba
Yosi            Tel-Aviv
dtype: object

In [93]:
dudes_cities.loc['Liran']

'Tal-Aviv'

In [94]:
dudes_cities.loc[ ['Liran', 'Nir', 'Guy'] ]

Liran         Tal-Aviv
Nir           Tel-Aviv
Nir         Beer-Sheba
Guy      Rishon LeZion
dtype: object

In [95]:
# Range (slice) by label also works, but ONLY if the index values are unique.
dudes_cities.loc['Liran':'Maor'] # slice depend on order, the index inclusive

Liran           Tal-Aviv
Oren       Rishon-Lezion
Ron            Jerusalem
Shachar     Petach-Tikva
Shalom      Petach Tikva
Tal        Rishon-LeZion
Yosi             Telaviv
Tomer             Beer-7
Nir             Tel-Aviv
Yaniv          Jerusalem
Maor               Haifa
dtype: object

In [96]:
dudes_cities.loc['Liran':'Ari']

Series([], dtype: object)

In [97]:
# dudes_cities.loc['Liran':'Nir']

We specifically note the `loc` option for **Boolean indexing**, which is similar to SQL's `WHERE` statement.

### _iloc_ - Integer-location-based

see more at [Selection by Position][1]

[1]: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#indexing-integer

In [98]:
dudes_cities

Alon          Beer Sheba
Ari             Tel Aviv
Barak              Haifa
Itamar     Rishon LeZion
Yarden         Beersheba
Ofer            Herzliya
Liran           Tal-Aviv
Oren       Rishon-Lezion
Ron            Jerusalem
Shachar     Petach-Tikva
Shalom      Petach Tikva
Tal        Rishon-LeZion
Yosi             Telaviv
Tomer             Beer-7
Nir             Tel-Aviv
Yaniv          Jerusalem
Maor               Haifa
Guy        Rishon LeZion
Nir           Beer-Sheba
Yosi            Tel-Aviv
dtype: object

In [99]:
dudes_cities.iloc[4] 

'Beersheba'

In [100]:
dudes_cities.iloc[ [0, 4, 1] ]

Alon      Beer Sheba
Yarden     Beersheba
Ari         Tel Aviv
dtype: object

In [101]:
dudes_cities.iloc[2:6] # location based slicing

Barak             Haifa
Itamar    Rishon LeZion
Yarden        Beersheba
Ofer           Herzliya
dtype: object

## Important attributes

`s.index` returns the index of `s` (instance of the `Index` class).

In [102]:
dudes_cities.index

Index(['Alon', 'Ari', 'Barak', 'Itamar', 'Yarden', 'Ofer', 'Liran', 'Oren',
       'Ron', 'Shachar', 'Shalom', 'Tal', 'Yosi', 'Tomer', 'Nir', 'Yaniv',
       'Maor', 'Guy', 'Nir', 'Yosi'],
      dtype='object')

In [103]:
for value in dudes_cities.index:
  if value.lower().startswith("a"): print(">>>", value)

>>> Alon
>>> Ari


`s.values` returns the values of `s` as an instance of the `array` class.

In [104]:
dudes_cities.values
# print(type(dudes_cities.values))

array(['Beer Sheba', 'Tel Aviv', 'Haifa', 'Rishon LeZion', 'Beersheba',
       'Herzliya', 'Tal-Aviv', 'Rishon-Lezion', 'Jerusalem',
       'Petach-Tikva', 'Petach Tikva', 'Rishon-LeZion', 'Telaviv',
       'Beer-7', 'Tel-Aviv', 'Jerusalem', 'Haifa', 'Rishon LeZion',
       'Beer-Sheba', 'Tel-Aviv'], dtype=object)

## Common operations

`Series` objects support nearly all the methods of numpy arrays, but also many more, including the following.

### Number of elements

Naturally, the built-in `len()` returns the number of elements in the Series. Note that the method `count()` does not count the nulls.

In [105]:
dudes_cities.count()    # <--- not including nulls / NaNs

20

In [106]:
len(dudes_cities) # use to count values in series

20

In [107]:
dudes_cities.size

20

In [108]:
dudes_cities.shape

(20,)

### `head()` and `tail()`

`head(n)` shows the top `n` records (n=5 by default)

In [109]:
dudes_cities.head()

Alon         Beer Sheba
Ari            Tel Aviv
Barak             Haifa
Itamar    Rishon LeZion
Yarden        Beersheba
dtype: object

`tail(n)` shows the bottom *n* records (n=5 by default)

In [110]:
dudes_cities.tail(3)

Guy     Rishon LeZion
Nir        Beer-Sheba
Yosi         Tel-Aviv
dtype: object

`sample(n)` shows *n* random records.

In [111]:
dudes_cities.sample(3) # use for introducing series

Itamar    Rishon LeZion
Tomer            Beer-7
Maor              Haifa
dtype: object

### `unique()` and `nunique()`

The `unique()` method returns an array of the unique elements of the calling object. The method `nunique()` returns the number of unique values in the series.

In [112]:
dudes_cities.unique() # use to introduce categorical values

array(['Beer Sheba', 'Tel Aviv', 'Haifa', 'Rishon LeZion', 'Beersheba',
       'Herzliya', 'Tal-Aviv', 'Rishon-Lezion', 'Jerusalem',
       'Petach-Tikva', 'Petach Tikva', 'Rishon-LeZion', 'Telaviv',
       'Beer-7', 'Tel-Aviv', 'Beer-Sheba'], dtype=object)

In [113]:
dudes_cities.index.unique() 

Index(['Alon', 'Ari', 'Barak', 'Itamar', 'Yarden', 'Ofer', 'Liran', 'Oren',
       'Ron', 'Shachar', 'Shalom', 'Tal', 'Yosi', 'Tomer', 'Nir', 'Yaniv',
       'Maor', 'Guy'],
      dtype='object')

In [114]:
dudes_cities.nunique() # use to introduce number of categorical values

16

### Math / statistics operations

In [115]:
dudes_ages

Avi    25
Ben    26
Gad    21
Ben    25
Hen    20
dtype: int64

In [116]:
dudes_ages.sum()

117

In [117]:
pd.Series(["red", "gold", "azul"]).sum()

'redgoldazul'

In [118]:
print(dudes_ages.min())
dudes_ages.max()

20


26

In [119]:
dudes_ages.mean()

23.4

In [120]:
dudes_ages.std()

2.701851217221259

### `value_counts()`

The [`value_counts()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html) method returns a new series in which the unique elements of the calling object are the index, and the values are their corresponding counts. The returned series is sorted by the counts.

In [121]:
dudes_cities

Alon          Beer Sheba
Ari             Tel Aviv
Barak              Haifa
Itamar     Rishon LeZion
Yarden         Beersheba
Ofer            Herzliya
Liran           Tal-Aviv
Oren       Rishon-Lezion
Ron            Jerusalem
Shachar     Petach-Tikva
Shalom      Petach Tikva
Tal        Rishon-LeZion
Yosi             Telaviv
Tomer             Beer-7
Nir             Tel-Aviv
Yaniv          Jerusalem
Maor               Haifa
Guy        Rishon LeZion
Nir           Beer-Sheba
Yosi            Tel-Aviv
dtype: object

In [122]:
dudes_cities.value_counts() # values distribution

Haifa            2
Rishon LeZion    2
Jerusalem        2
Tel-Aviv         2
Beer Sheba       1
Tel Aviv         1
Beersheba        1
Herzliya         1
Tal-Aviv         1
Rishon-Lezion    1
Petach-Tikva     1
Petach Tikva     1
Rishon-LeZion    1
Telaviv          1
Beer-7           1
Beer-Sheba       1
dtype: int64

In [123]:
dudes_cities.value_counts(normalize=True) # distribution ratio

Haifa            0.10
Rishon LeZion    0.10
Jerusalem        0.10
Tel-Aviv         0.10
Beer Sheba       0.05
Tel Aviv         0.05
Beersheba        0.05
Herzliya         0.05
Tal-Aviv         0.05
Rishon-Lezion    0.05
Petach-Tikva     0.05
Petach Tikva     0.05
Rishon-LeZion    0.05
Telaviv          0.05
Beer-7           0.05
Beer-Sheba       0.05
dtype: float64

In [124]:
# What does the following code do? Percentage of 3 most frequent values

ser_colors = pd.Series(["green", "pink", "white", "green", "blue", "blue", "green", "yellow", "pink", "green", "gold"])

# ser_colors.value_counts(normalize=True).iloc[:3].sum()

In [125]:
(
  ser_colors
  .value_counts(normalize=True)
  .iloc[:3]
  .sum()
)

0.7272727272727273

### `replace()`

The [`replace(to_replace, value)`][1] will replace values given in `to_replace` with `value`. We can pass `replace()` a dictionary indicating that dict's keys should be replaced by the dict's values.

[1]: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.replace.html

> **Note:** Like many other functions, `replace()` accepts a boolean argument **`inplace`** indicating whether to perform the operation in place or not. This kwarg affects the return value.

In [126]:
cities_map_dict = {
    'Tel Aviv': 'Tel-Aviv',
    'Rishon LeZion': 'Rishon-LeZion',
    'Petah Tikva': 'Petah-Tikva',
    'Kfar Saba': 'Kfar-Saba',
    'Beersheba': 'Beer-Sheba',
    'Beer Sheba': 'Beer-Sheba',
    'Rishon Lezion': 'Rishon-LeZion',
    "Be'er-Sheva": "Beer-Sheba",
    'Telaviv': 'Tel-Aviv',
    'Beer-7': 'Beer-Sheba',
    'Rishon-Lezion': 'Rishon-LeZion',
    'Petach Tikva': 'Petach-Tikva',
    'Tal-Aviv': 'Tel-Aviv'
}

In [127]:
dudes_cities2 = dudes_cities.replace(cities_map_dict) get dictionary - old values: new values

SyntaxError: invalid syntax (Temp/ipykernel_10280/2531889518.py, line 1)

In [ ]:
dudes_cities2

In [ ]:
dudes_cities2.value_counts(normalize=True)

> **Your turn:** <--- Homework 16/7/2024

Given the *name_ser* series below:

1. How many values are there in the series?
2. How many times does each name appear in the series?
3. Correct the series so that the names 'james' and 'frank' are capitalized (upper-cased).
4. How many times do each name appear in the series now? Display the percentage instead of the count.
5. Print the most common name (pick one, if there is more than one).
6. Print the least common name (pick one, if there is more than one).

In [ ]:
names = ['John','William','James','Charles','George',
         'frank','Joseph','Thomas','james','Charles',
         'Arthur','Albert','William','Charles','Frank',
         'Charles','Sam','Jacob','Elmer','Arthur','Arthur']

names_ser = pd.Series(data=names)
names_ser

In [ ]:
# 1. How many values are there in the series?
names_ser.count()
# names_ser.size

In [ ]:
# 2. How many times does each name appear in the series?
names_ser.value_counts()
# names_ser.value_counts()

In [ ]:
# 3. Correct the series so that the names 'james' and 'frank' are capitalized (upper-cased).
correct_values = {
    "frank" : "Frank",
    "james" : "James"
}

names_ser.replace(correct_values, inplace =True)



'''
corrected_names = {
    "frank": "Frank",
    "james": "James"
}
# Either use inplace=True or re-assign to names_ser (names_ser = names_ser.replace...) for the solution to #4 to work.
names_ser.replace(corrected_names, inplace=True)
names_ser

'''

In [ ]:
print(names_ser)

In [ ]:
# 4. How many times do each name appear in the series now? Display the percentage instead of the count.
names_ser.value_counts(normalize = True) 

In [ ]:
# 5. Print the most common name (pick one, if there is more than one).
count_name_series = names_ser.value_counts()
count_name_series.index[0]

# names_ser.value_counts().index[0]


In [ ]:
# 6. Print the least common name (pick one, if there is more than one).
count_name_series = names_ser.value_counts()
count_name_series.index[-1]

# names_ser.value_counts(ascending=True).index[0]

### Solution

In [ ]:
# 1.
names_ser.size

In [ ]:
# 2.
names_ser.value_counts()

In [ ]:
# 3.
corrected_names = {
    "frank": "Frank",
    "james": "James"
}
# Either use inplace=True or re-assign to names_ser (names_ser = names_ser.replace...) for the solution to #4 to work.
names_ser.replace(corrected_names, inplace=True)
names_ser

In [ ]:
# 4.
names_ser.value_counts(normalize=True)

In [ ]:
# 5.
names_ser.value_counts().index[0]

In [ ]:
# 6.
names_ser.value_counts(ascending=True).index[0]

### _apply(func)_

In [ ]:
dudes_cities

The [`apply(func)`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.apply.html) method applies the function `func` on each element of the calling series and returns a new series with the returned values.

In [ ]:
def first_letter(value):
    return value[::-1]

In [ ]:
dudes_cities.apply(first_letter) #.head(10)

In [ ]:
pd.Series(dudes_cities.index).apply(lambda x: x.lower())

In [ ]:
dudes_cities.index

In [ ]:
dudes_cities.apply(len)  #.head()

In [ ]:
len(dudes_cities)

### `sort_values()`

The [`sort_values()`][1] method sorts the elements of the calling object **and keeps their index**.

[1]: http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.sort_values.html#pandas.Series.sort_values "Series.sort_values() documentation"

In [ ]:
dudes_cities2

In [ ]:
dudes_cities2.value_counts(ascending=True) #<- default value

In [ ]:
ser_colors

In [ ]:
ser_colors.sort_values(ascending=False)

In [ ]:
dudes_cities2.sort_index()

In [ ]:
dudes_cities2.sort_values(ascending=False, inplace=True)
dudes_cities2

#### [optional] The `key` function

We note that the `key` argument does not work element-wise (like in the built-in `sorted()`), but rather in a *vectorized* way. This means `key` expects a Series object and returns an array-like. In other words, the key function returns the sorting criteria for *all* the values to be sorted.

In [ ]:
dudes_cities.sort_values(key=lambda ser_values: ser_values.str.len())

And another example to illustrate a more complicated functions.

In [ ]:
def num_of_words(city):
    num_words = max(len(city.split('-')), len(city.split()))
    print(f">>> {city:15s} -> {num_words}")
    return num_words

def num_of_words_per_value(ser_values):
    ser_sort_criteria = ser_values.apply(num_of_words)
    print("### sort criteria series:")
    print(ser_sort_criteria)
    return ser_sort_criteria

dudes_cities.sort_values(key=num_of_words_per_value)

### The _str_ accessor

When the series dtype is 'str', there is a collection of special methods available to work with strings. They are all collected and accessed through the accessor [_str_][1], and are called only through it.

[1]: http://pandas.pydata.org/pandas-docs/stable/api.html#string-handling "The str accessor"

In [ ]:
dudes_cities2 = dudes_cities2.str.lower()
dudes_cities2

In [ ]:
dudes_cities.str.upper() # use accesor string to prevent ambigiouty

In [ ]:
dudes_cities.str.upper() #.head()

In [ ]:
dudes_cities

In [ ]:
dudes_cities.str.replace('Sheba', 'Sheva').head()

> **Note:** _pandas_ supports several other accessors, including [`plot`][3] for plotting, [`dt`][1] for datetime dtype and [`cat`][2] for categorical data.

[1]: http://pandas.pydata.org/pandas-docs/stable/api.html#datetimelike-properties "The dt accessor"
[2]: http://pandas.pydata.org/pandas-docs/stable/api.html#categorical "The cat accessor"
[3]: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.plot.html "The plot accessor"

## Boolean Masking

Boolean masks are means for selecting values meeting a certain criteria from a Series. The boolean mask is generated by applying a binary test (True/False) to each value in the series, producing a new series of True/False values with the exact same index values. The boolean mask can then be used to filter values from the originating series.

In [ ]:
# Boolean mask
dudes_cities2.str.endswith('n')

In [ ]:
# Using the boolean mask to filter values
dudes_cities2.loc[dudes_cities2.str.endswith('n')]

In [ ]:
dudes_ages

In [ ]:
(dudes_ages >= 21)

In [ ]:
dudes_ages.loc[dudes_ages >= 21]

In [ ]:
(dudes_ages >= 21) & (dudes_ages <=25)

In [ ]:
dudes_ages[(dudes_ages >= 21) & (dudes_ages <=25)]

In [ ]:
selection_criteria = (dudes_ages >= 21) & (dudes_ages <=25)
selection_criteria

logic operators &(AND) |(OR) ~()NOT  must be in ()

In [ ]:
dudes_ages[selection_criteria]

In [ ]:
dudes_ages

In [ ]:
dudes_ages[[True, False, True, False, True]]

In [ ]:
dudes_ages[dudes_ages.between(25, 27)]

### Missing values

It is possible to find missing values in a series. Missing values are **NaN** values, which are common when dealing with tabular data, where not all the fields in a row must have values (sometimes they must and we need to find those which don't and deal with them).

There are several related methods which can be used:<br/>
`isnull()`, `isna()`, `notnull()` and `notna()`. These methods return a new series made of boolean, one per each value (matching index is preserved) which is the result of applying the requested test (i.e., *isnull*, *notnull*, and so on).

In [ ]:
# pandas is on the top of numpy
ser_nan = pd.Series([100, 150, np.NaN, 133, 97, np.NaN, 12, np.NaN, 99])

In [ ]:
ser_nan.isnull() #.sum()

In [ ]:
ser_nan.isna()

In [ ]:
ser_nan.notnull().sum() # ser_nan.count()

In [ ]:
ser_nan.notna()

In [ ]:
ser_nan_not_null = ser_nan.loc[ser_nan.notnull()]
ser_nan_not_null

In [ ]:
ser_nan

In [ ]:
ser_nan[ser_nan.notnull()]

In [ ]:
~ser_nan.isnull() # ~ not

In [ ]:
# imputation - manage missing value
# if missing values < 10% 

ser_nan.mean()

In [ ]:
ser_nan.fillna(ser_nan.mean())

> **Your turn 1:**<br/>
> Using the *names_series* (defined below):<br/>
> 1. Convert all names to lower case.<br/>
> 2. Find all names which end with an 's'.<br/>
> 3. Sort the names lexicographically and print the top 10.<br/>
> 4. What is the most common name?
> 5. [hard] Which is the longest name? (print only one if there are multiple names with the same length).<br/>

In [ ]:
names = ['John','William','James','Charles','George',
         'Frank','Joseph','Thomas','James','Charles',
         'Arthur','Albert','William','Charles','Frank',
         'Charles','Sam','Jacob','Elmer','Arthur','Arthur']

names_series = pd.Series(data=names, index=names)
print(names_series)

In [ ]:
# 1. Convert all names to lower case.
names_series.str.lower()


In [ ]:
# 2. Find all names which end with an 's'.
names_series.str.endswith('s')

In [ ]:
# 3. Sort the names lexicographically and print the top 10.
names_series.sort_values().head(10)

In [ ]:
# 4. What is the most common name?

names_series.value_counts()[0]

In [ ]:
# 5. [hard] Which is the longest name? (print only one if there are multiple names with the same length).
names_series_len = names_series.str.len().sort_values(ascending=False)
print(names_series_len.index[0], names_series_len[0]) 

In [ ]:
# optional solution
names_series[names_series.str.len() == names_series.str.len().max()].iloc[0]

### Solution

In [ ]:
# 1. Convert all names to lower case.
names_series = names_series.str.lower()
names_series

In [128]:
# 2. Find all names which end with an 's'.
names_series.loc[names_series.str.endswith("s")]

NameError: name 'names_series' is not defined

In [ ]:
# 3. Sort the names lexicographically and print the top 10.
names_series.sort_values().iloc[:10]
# Another option
# names_series.sort_values().head(10)

In [ ]:
# 4.What is the most common name?
names_series.value_counts().index[0]

In [ ]:
# 5. Which is the longest name? (print only one if there are multiple names with the same length).
names_series.str.len().sort_values(ascending=False).index[0]

In [ ]:
names_series.apply(len).sort_values(ascending=False).index[0]

> **Your turn 2:** <br/> 
Homework 18/7/2024

Import the data from the file *english words.txt* (available [here](https://drive.google.com/drive/folders/1KQXg5CpZ8u59ybkvOnFPzo_WaLHxek_g?usp=sharing)), using the command `pd.read_csv('english words.txt').squeeze=()` and follow the next questions and tasks:
>* Task 1: How many words are listed in the file? Consider that there may be NaNs among the words...
>* Task 2: Find the longest words in the file.
>* Task 3: Find all the words which are palindromes.

In [ ]:
words = pd.read_csv('english words.txt', header=None).squeeze()

In [ ]:
words.sample(5)

In [ ]:
# Task 1: How many words are listed in the file? Consider that there may be NaNs among the words...

words[words.isnull()] # How many NaNs in the set # words.isnull() is used as selection criteria

In [ ]:
# Words not null
len(words) - len(words[words.isnull()])

In [ ]:
# Words not null - optimal option
words.count()

In [ ]:
# remove the null values
words = words.loc[words.notnull()]
len(words)

In [ ]:
# Task 2: Find the longest words in the file.
words.loc[words.str.len() == words.str.len().max()]

In [ ]:
# Task 3: Find all the words which are palindromes.

def reverse_word(word):
    return word[::-1]

# Assuming words is a pandas Series
words_series = pd.Series(words)

# Ensure all elements are strings
words_series = words_series.astype(str)

# Create a series of reversed words
# words_reversed = words_series.apply(lambda word: word[::-1]) # shortcut way
words_reversed = words.apply(reverse_word)
words_reversed

In [ ]:
words == words_reversed # camparing element in serious - vector actions

In [ ]:
words[words == words_reversed]

In [ ]:
# Assuming words_series is your pandas Series
data_types = words_series.apply(lambda x: type(x))
print(data_types)

In [ ]:
# What if we wanted only 4 letters or longer palindromes:
(words == words_reversed) & (words.str.len() >= 4)

In [ ]:
words.loc[(words == words_reversed) & (words.str.len() >= 4)]

### Solution

In [ ]:
words.sample(5)

In [ ]:
# Task 1: How many words are listed in the file? Note there are some missing entries and clean them...
words[words.isnull()]

In [ ]:
len(words) - len(words[words.isnull()])

In [ ]:
# Another option
len(words.loc[~words.isnull()])
# len(words.loc[words.notnull()])

In [ ]:
# Yet another option
words.loc[words.notnull()].shape

In [ ]:
# And another
words.count()

In [ ]:
# Remove the null values
words = words.loc[words.notnull()]
len(words)

In [ ]:
# Task 2: Find the longest words in the file and

# The length of the longest word
words.str.len().max()

In [ ]:
# Boolean mask finding all the words of the "longest word" length
words.str.len() == words.str.len().max()

In [ ]:
# Retrieve only the longest words
words.loc[words.str.len() == words.str.len().max()]

In [ ]:
# Task 3: Find all the words which are palindromes.

# Create a series of reversed words
words_reversed = words.apply(lambda word: word[::-1])
words_reversed

In [ ]:
# Boolean mask for words which are the same as their reversed counterparts.
words == words_reversed

In [ ]:
words[words == words_reversed]

In [ ]:
# Option 2
words.loc[words == words.apply(lambda word: word[::-1])]

In [ ]:
# What if we wanted only 4 letters or longer palindromes?

(words == words_reversed) & (words.str.len() >= 4)

In [ ]:
words.loc[(words == words_reversed) & (words.str.len() >= 4)]

## @@@ Extras

In [ ]:
# Binning - categorizing numeric value ranges
customer_ages = pd.Series([18, 27, 83, 41, 15, 38, 62, 77, 25, 34, 46])
# Function cut does not depend on series
customer_age_groups = pd.cut(
    customer_ages,
    bins=[0, 18, 30, 60, 120],
    labels=["underage", "young adult", "adult", "senior"]
)
customer_age_groups

In [129]:
# Panda's "in" operator - look for a value in a list of other values
can_sell_to = customer_age_groups.isin(["young adult", "adult", "senior"])
can_sell_to

NameError: name 'customer_age_groups' is not defined

In [ ]:
customer_age_groups[can_sell_to]

# The _DataFrame_ class

## Basics

The _DataFrame_ is the primary _pandas_ data structure, and can be thought of as a dict-like container for _Series_ objects. _DataFrame_ objects have many attributes and support many methods, which are all documented alphabetically [here][1] and by subject [here][2].

Like in the _Series_ objects, the rows are labeled by the **_index_**, but for DataFrames there is another labeling for the columns, called **columns** (which is also an instance of the _Index_ class). If either is not specified, then the elements are labeled by the regular row-counter.

Most of the fundamental functionalities of _DataFrames_ are similar to those of the _Series_, so we will suffice with a short demonstration.

[1]: http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html#pandas.DataFrame "DataFrame API documentation A-Z"
[2]: http://pandas.pydata.org/pandas-docs/stable/api.html#dataframe "DataFrame API documentation"

In [ ]:
dudes_names = pd.Series(['Avi', 'Ben', 'Avi', 'Dor', 'Hen'])
dudes_ages = pd.Series([32, 34, 21, 53, 19])
dudes_married = pd.Series([True, False, False, True, True])

In [ ]:
data = {'Name': dudes_names,
        'Age': dudes_ages,
        'is_married': dudes_married}

In [ ]:
df = pd.DataFrame(data)
df.head()

In [ ]:
df.columns

In [ ]:
df.index

In [ ]:
df * 2

In [ ]:
df["Name"] # defalut index of column

In [ ]:
df.Name

In [130]:
2024 - df['Age'] # 

NameError: name 'df' is not defined

In [ ]:
df['birth year'] = 2024 - df['Age'] # calculated column # create new column

In [ ]:
df

In [ ]:
df["birth year"] # the result in format series

In [ ]:
~df['is_married']

In [131]:
df['is_single'] = ~df['is_married']

NameError: name 'df' is not defined

In [ ]:
df

In [ ]:
df["new_age"] = df.Age / 2 # new column

In [ ]:
df['Age'] + df['new_age']

In [ ]:
df['age+new_age'] = df.Age + df['new_age']

In [132]:
df["Age"]

NameError: name 'df' is not defined

In [ ]:
df["stam"] = [111, 222, 333, 444, 555]

In [ ]:
df

In [ ]:
(df['Age'] > 23) & (df['new_age'] < 100)

In [ ]:
df[(df['Age'] > 23) & (df['new_age'] < 100)] # bolean masks in data frame results rows
# & -> and
# | -> or
# =
# != -> not =

> **Note:** Most of the time you will not have to construct a DataFrame yourself, rather it will be uploaded from some data resource, e.g. file or database.

Adding columns is as simple as adding an item to a dictionary...

In [ ]:
df['birth year'] = 2021 - df['Age']
df

### Axis

In [133]:
dfx = pd.DataFrame(
    [
        [1,2,3],
        [10, 20, 30],
        [100, 200, 300]
    ]
)
dfx

0    1    2
0    1    2    3
1   10   20   30
2  100  200  300

In [134]:
# Axis 0 - operate on the COLUMNS
dfx.sum(axis=0) # default, agregate values in columns

0    111
1    222
2    333
dtype: int64

In [135]:
# Axis 1 - operate on the ROWS
dfx.sum(axis=1) # agregate values in rows

0      6
1     60
2    600
dtype: int64

## Important attributes

Note that just like the intuitive rows index is an `Index` instance, the `columns` attribute is also an `Index` instance. If we choose a single row, we get a `Series` object whose index is `df.columns`.

In [136]:
df.index

NameError: name 'df' is not defined

In [ ]:
df.columns

In [ ]:
df.shape # the result is tuple with 2 values

## Indexing and selecting data

In [ ]:
df.Age

In [ ]:
# Select a subset of columns
df[["Age","is_married"]] # list of columns result new data frame with selected columns

In [ ]:
# Select rows by index values range
df.loc[1:3]

In [ ]:
# Filter by: rows index range and a single column
df.loc[1:3, "Age"] # the result is series

In [ ]:
# Filter by: rows index range and multiple columns
df.loc[1:3, ["Age", "is_married"]]

In [137]:
# Filter by: arbitrary rows indexes and multiple columns
df.loc[[2, 0, 3], ["Age", "is_married"]] # this index so it's inclusive

NameError: name 'df' is not defined

In [ ]:
# Filter by: rows index range and columns range
df.loc[:3, "Age":"is_single"]

Selection by location using `.iloc` is also available.

In [ ]:
df.iloc[2:4, [0, 2]] # iloc is for location, so it exclusive

## Common operations

DataFrames have [many methods](https://pandas.pydata.org/pandas-docs/stable/reference/frame.html), which are for many purposes. It is useful to think of them in "families", as grouped in the docs.

### `info()` and `describe()` (DataFrame overview)

In [ ]:
df.info() # metadata about dataframe

In [138]:
df.describe() # descriptive statistics

NameError: name 'df' is not defined

### `head()` and `tail()` (Indexing and iteration)

The methods _head()_ and _tail()_ print the first and last rows respectively of a DataFrame.

In [ ]:
df.head()

In [139]:
df.tail(3)

NameError: name 'df' is not defined

In [ ]:
df.sample(3) # default is 1

### `set_index()` and `reset_index()` (Reindexing / selection / label manipulation)

`set_index()` allows to set the `DataFrame` index (row labels) using one or more existing columns.

In [ ]:
df.head()

In [ ]:
df.set_index('Name', inplace=True, drop=True) # the coulmn 'Name' sets as index # index does not have to be unique

In [140]:
df.loc[["Ben", "Avi"], ["Age", "is_single"]]

NameError: name 'df' is not defined

*reset_index()* reset the index, turn the index values to a column and generate a sequence of integers as index

In [ ]:
df.reset_index(inplace=True) # undo to original index
df

### `read_csv()`

One of `pandas` greatest features is its wrappers for I/O operations with various file formats. The entire list is documented [here][1], but we will look here the most common method - [`read_csv()`][2].

[1]: http://pandas.pydata.org/pandas-docs/stable/api.html#input-output "DataFrame I/O documentation"
[2]: http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html "read_csv() documentation"

> **Note:** For this example to run, the file 'dessert.csv' (available [here](https://drive.google.com/drive/folders/1KQXg5CpZ8u59ybkvOnFPzo_WaLHxek_g?usp=sharing)) should be uploaded to your local environment. If you run this notebook in Google Colab you may use the Upload option.

In [ ]:
!head dessert.csv

In [141]:
pd.read_csv?

In [142]:
df = pd.read_csv("dessert.csv", index_col='id')
df.head()

day.of.week  num.of.guests        hour  table  dessert
id                                                       
1       Monday              2     evening     13     True
2     Saturday              4       night     19     True
3     Saturday              4  after-noon     12    False
4     Thursday              3       night      5     True
5       Monday              7        noon     10     True

In [143]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1000 entries, 1 to 1000
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   day.of.week    1000 non-null   object
 1   num.of.guests  1000 non-null   int64 
 2   hour           1000 non-null   object
 3   table          1000 non-null   int64 
 4   dessert        1000 non-null   bool  
dtypes: bool(1), int64(2), object(2)
memory usage: 40.0+ KB


> **NOTE:** `read_csv()` is a very powerful method. Note that many things are inferred automatically. In our example the header, the dtypes and the fact that the 'dessert' column is Boolean.

### `apply()` (Function application, GroupBy & window)

This method applies a function on every **row (or column)** of the data frame. The objects that are passed to the function (input argument) are Series objects (either a column or a row).

The API is documented [here][1] and a detailed explanation on its usage ca be found [here][2].

[1]: http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.apply.html "DataFrame.apply() API"
[2]: http://pandas.pydata.org/pandas-docs/stable/basics.html#row-or-column-wise-function-application "DataFrame.apply() documentation"

In [144]:
df.head()

day.of.week  num.of.guests        hour  table  dessert
id                                                       
1       Monday              2     evening     13     True
2     Saturday              4       night     19     True
3     Saturday              4  after-noon     12    False
4     Thursday              3       night      5     True
5       Monday              7        noon     10     True

By default, the given function is applied on each column.

In [145]:
df.apply?

In [146]:
df.apply(min, axis=0) # default, agregation per column

day.of.week          Friday
num.of.guests             1
hour             after-noon
table                     1
dessert               False
dtype: object

In [147]:
df.apply(min, axis=0) # default agregation per column, 1st argument is function's name

day.of.week          Friday
num.of.guests             1
hour             after-noon
table                     1
dessert               False
dtype: object

As we saw above, when selecting a single row, we get a Series object which its Index is the `columns` attribute of the data frame.
If we apply a function on the rows (using `axis=1`), we sometimes need to take it into account.

In [148]:
def is_weekend(row):

    return row.loc['day.of.week'] in ['Friday', 'Saturday']
    # same as:
    # return row['day.of.week'] in ['Friday', 'Saturday']

In [149]:
df['is_weekend'] = df.apply(is_weekend, axis=1)  # agregation per row, apply is_weekend on each row --> axis=1

In [150]:
df.head()

day.of.week  num.of.guests        hour  table  dessert  is_weekend
id                                                                   
1       Monday              2     evening     13     True       False
2     Saturday              4       night     19     True        True
3     Saturday              4  after-noon     12    False        True
4     Thursday              3       night      5     True       False
5       Monday              7        noon     10     True       False

> **Note:** `apply()` is the most common, but there are also `map()` and `applymap()`. On the difference between them you can read in [this SO answer](https://stackoverflow.com/a/19798528/3121900).

### `drop()` (Reindexing / selection / label manipulation)

The method [`drop()`][1] removes the specified rows/columns from the calling dataframe.

[1]: http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html "drop() API"

In [151]:
df.head()

day.of.week  num.of.guests        hour  table  dessert  is_weekend
id                                                                   
1       Monday              2     evening     13     True       False
2     Saturday              4       night     19     True        True
3     Saturday              4  after-noon     12    False        True
4     Thursday              3       night      5     True       False
5       Monday              7        noon     10     True       False

In [152]:
df[~(df['day.of.week'] == 'Monday')]

day.of.week  num.of.guests        hour  table  dessert  is_weekend
id                                                                     
2       Saturday              4       night     19     True        True
3       Saturday              4  after-noon     12    False        True
4       Thursday              3       night      5     True       False
6         Friday              2     evening     19    False        True
8         Sunday              4        noon      4     True       False
...          ...            ...         ...    ...      ...         ...
996       Friday              3  after-noon      5    False        True
997    Wednesday              5     evening     14     True       False
998     Thursday              2     evening     20     True       False
999     Saturday              2     evening      3     True        True
1000      Friday              2     evening      9    False        True

[895 rows x 6 columns]

In [153]:
# Remove rows (by indexes rows and columns)
df.drop(index=[2, 3])

day.of.week  num.of.guests        hour  table  dessert  is_weekend
id                                                                     
1         Monday              2     evening     13     True       False
4       Thursday              3       night      5     True       False
5         Monday              7        noon     10     True       False
6         Friday              2     evening     19    False        True
7         Monday              2     evening     22     True       False
...          ...            ...         ...    ...      ...         ...
996       Friday              3  after-noon      5    False        True
997    Wednesday              5     evening     14     True       False
998     Thursday              2     evening     20     True       False
999     Saturday              2     evening      3     True        True
1000      Friday              2     evening      9    False        True

[998 rows x 6 columns]

In [154]:
# Remove columns (by column name / index) 

In [155]:
df.head(3)

day.of.week  num.of.guests        hour  table  dessert  is_weekend
id                                                                   
1       Monday              2     evening     13     True       False
2     Saturday              4       night     19     True        True
3     Saturday              4  after-noon     12    False        True

In [156]:
df.sort_values(by=["day.of.week", "num.of.guests"], ascending=[True, False]) # sort values with numbers of criteria

day.of.week  num.of.guests     hour  table  dessert  is_weekend
id                                                                 
514      Friday             10    night      3     True        True
525      Friday             10  evening     18    False        True
559      Friday             10     noon     20     True        True
844      Friday              9     noon     12     True        True
986      Friday              9  evening     16    False        True
..          ...            ...      ...    ...      ...         ...
616   Wednesday              1  evening     21    False       False
672   Wednesday              1     noon     20     True       False
835   Wednesday              1  evening     16    False       False
864   Wednesday              1    night     12     True       False
914   Wednesday              1     noon     23     True       False

[1000 rows x 6 columns]

In [157]:
df.rename(columns={"hour": "hour-of-day"}) # renaming indexing

day.of.week  num.of.guests hour-of-day  table  dessert  is_weekend
id                                                                     
1         Monday              2     evening     13     True       False
2       Saturday              4       night     19     True        True
3       Saturday              4  after-noon     12    False        True
4       Thursday              3       night      5     True       False
5         Monday              7        noon     10     True       False
...          ...            ...         ...    ...      ...         ...
996       Friday              3  after-noon      5    False        True
997    Wednesday              5     evening     14     True       False
998     Thursday              2     evening     20     True       False
999     Saturday              2     evening      3     True        True
1000      Friday              2     evening      9    False        True

[1000 rows x 6 columns]

> **Your turn:** Do the following with the dessert dataset:
> * Add a column called `not_dessert` with the inverse of `dessert`
> * Add a column with random positive numbers and call it `Bill`
    * You can use random.randint (make sure to import the `random` module first)
> * Use `sort_values()` to sort the data by the bill
> * Use `rename()` to change all the points (.) to underscores (_)
> * Create separate dataframes for working days and weekends

In [158]:
df = pd.read_csv("dessert.csv", index_col='id')

In [159]:
df.sample(10)

day.of.week  num.of.guests        hour  table  dessert
id                                                        
614     Tuesday              3     evening      3     True
405      Monday              3       night     16     True
314      Monday              4        noon      7     True
195      Sunday              4       night     21     True
468   Wednesday              2     evening     20     True
368      Friday              4     evening     20    False
918      Friday              3       night     22    False
772     Tuesday              7        noon     18     True
441      Sunday              2  after-noon      8    False
312      Friday              2     evening     17    False

In [160]:
1. # Add a column called not_dessert with the inverse of dessert
df['not_dessert'] = ~df['dessert']
df.head()

day.of.week  num.of.guests        hour  table  dessert  not_dessert
id                                                                    
1       Monday              2     evening     13     True        False
2     Saturday              4       night     19     True        False
3     Saturday              4  after-noon     12    False         True
4     Thursday              3       night      5     True        False
5       Monday              7        noon     10     True        False

In [161]:
2. # Add a column with random positive numbers and call it Bill
# You can use random.randint (make sure to import the random module first)
import random

df['Bill'] = np.random.randint(1, 101, size=len(df))  # Random integers between 1 and 100
df


day.of.week  num.of.guests        hour  table  dessert  not_dessert  Bill
id                                                                            
1         Monday              2     evening     13     True        False    76
2       Saturday              4       night     19     True        False    64
3       Saturday              4  after-noon     12    False         True    47
4       Thursday              3       night      5     True        False    20
5         Monday              7        noon     10     True        False    81
...          ...            ...         ...    ...      ...          ...   ...
996       Friday              3  after-noon      5    False         True    14
997    Wednesday              5     evening     14     True        False    79
998     Thursday              2     evening     20     True        False    32
999     Saturday              2     evening      3     True        False    40
1000      Friday              2     evening      9    False         True    45

[1000 rows x 7 columns]

In [162]:
# 3. Use sort_values() to sort the data by the Bill

df.sort_values(by="Bill", inplace=True)
df 

day.of.week  num.of.guests        hour  table  dessert  not_dessert  Bill
id                                                                           
754     Tuesday              4       night      4     True        False     1
105      Sunday              2     evening      3     True        False     1
872    Saturday              9        noon     13     True        False     2
916      Friday              4  after-noon     18    False         True     2
462      Friday              3     evening     21    False         True     2
..          ...            ...         ...    ...      ...          ...   ...
186      Monday              6     evening     13     True        False   100
103    Saturday              2        noon      1     True        False   100
338     Tuesday              4     evening     19     True        False   100
73       Monday              2       night      1     True        False   100
771   Wednesday              3     evening      6     True        False   100

[1000 rows x 7 columns]

In [163]:
# 4. Use rename() to change all the points (.) to underscores (_)
df.rename(columns={
    "day.of.week": "day_of_week",
    "num.of.guests":"num_of_guests"},
          inplace=True) 
#df.rename(mapper={"day.of.week": "day_of_week", "num.of.guests": "num_of_guests"}, axis=1, inplace=True)
df.head()

day_of_week  num_of_guests        hour  table  dessert  not_dessert  Bill
id                                                                           
754     Tuesday              4       night      4     True        False     1
105      Sunday              2     evening      3     True        False     1
872    Saturday              9        noon     13     True        False     2
916      Friday              4  after-noon     18    False         True     2
462      Friday              3     evening     21    False         True     2

In [164]:
# 5. Create separate dataframes for working days and weekends

df_weekends = df.loc[df['day_of_week'].isin(['Saturday', 'Sunday'])]
df_weekends.head()

day_of_week  num_of_guests        hour  table  dessert  not_dessert  Bill
id                                                                           
105      Sunday              2     evening      3     True        False     1
872    Saturday              9        noon     13     True        False     2
814    Saturday              5        noon      3     True        False     3
400    Saturday              3     evening      5    False         True     3
508    Saturday              5  after-noon      8    False         True     3

In [165]:
df_work_days = df.loc[~df['day_of_week'].isin(['Saturday', 'Sunday'])]
df_work_days.head()

# df_weekend = df[(df.day_of_week == "Friday") | (df.day_of_week == "Saturday")]  # df[df.day_of_week.isin(["Friday", "Saturday"])]
# df_weekend.head()

day_of_week  num_of_guests        hour  table  dessert  not_dessert  Bill
id                                                                           
754     Tuesday              4       night      4     True        False     1
916      Friday              4  after-noon     18    False         True     2
462      Friday              3     evening     21    False         True     2
719      Friday              2  after-noon      4    False         True     2
958      Monday             10     evening     17     True        False     2

### Solution

In [166]:
# Add a column called not_dessert with the inverse of dessert
df["not_dessert"] = ~df["dessert"]
df.head()

day_of_week  num_of_guests        hour  table  dessert  not_dessert  Bill
id                                                                           
754     Tuesday              4       night      4     True        False     1
105      Sunday              2     evening      3     True        False     1
872    Saturday              9        noon     13     True        False     2
916      Friday              4  after-noon     18    False         True     2
462      Friday              3     evening     21    False         True     2

In [167]:
# Add a column with random positive numbers and call it Bill
from random import randint

df["bill"] = None
df.bill = df.bill.apply(lambda _: randint(5, 200))
df.head()

# Alternatively (using simple for loops)
random_bills = []
for _ in range(len(df)):
  random_bills.append(
      randint(5, 200)
  )
df["bill"] = random_bills

# And yet another alternative
# df["bill"] = np.random.random_integers(5, 200, size=len(df))

In [168]:
# Use sort_values() to sort the data by the bill
df.sort_values(by="bill")

day_of_week  num_of_guests        hour  table  dessert  not_dessert  Bill  \
id                                                                              
398    Thursday              5     evening     21    False         True    38   
846     Tuesday              4  after-noon     13    False         True    68   
347    Saturday              2       night      4     True        False    35   
278    Saturday              2  after-noon     14    False         True    98   
842      Sunday              5       night     19     True        False    47   
..          ...            ...         ...    ...      ...          ...   ...   
716    Saturday              2     evening      7     True        False    81   
437      Monday              3  after-noon      3    False         True    29   
721   Wednesday              4        noon     12     True        False    67   
338     Tuesday              4     evening     19     True        False   100   
727      Monday              3  after-noon      8    False         True    39   

     bill  
id         
398     5  
846     5  
347     5  
278     5  
842     5  
..    ...  
716   200  
437   200  
721   200  
338   200  
727   200  

[1000 rows x 8 columns]

In [169]:
# Use rename() to change all the points (.) to underscores (_)
df.rename(mapper={"day.of.week": "day_of_week", "num.of.guests": "num_of_guests"}, axis=1, inplace=True)
df.head()

day_of_week  num_of_guests        hour  table  dessert  not_dessert  Bill  \
id                                                                              
754     Tuesday              4       night      4     True        False     1   
105      Sunday              2     evening      3     True        False     1   
872    Saturday              9        noon     13     True        False     2   
916      Friday              4  after-noon     18    False         True     2   
462      Friday              3     evening     21    False         True     2   

     bill  
id         
754    24  
105    60  
872   148  
916   169  
462    75

In [170]:
# [for those interested...] Generic rename

rename_dict = {}
for column_name in df.columns:
    if "." in column_name:
        rename_dict[column_name] = column_name.replace(".", "_")

df.rename(columns=rename_dict, inplace=True)
df.head()

day_of_week  num_of_guests        hour  table  dessert  not_dessert  Bill  \
id                                                                              
754     Tuesday              4       night      4     True        False     1   
105      Sunday              2     evening      3     True        False     1   
872    Saturday              9        noon     13     True        False     2   
916      Friday              4  after-noon     18    False         True     2   
462      Friday              3     evening     21    False         True     2   

     bill  
id         
754    24  
105    60  
872   148  
916   169  
462    75

In [171]:
# Create separate dataframes for working days and weekends
df_weekend = df[(df.day_of_week == "Friday") | (df.day_of_week == "Saturday")]  # df[df.day_of_week.isin(["Friday", "Saturday"])]
df_weekend.head()

day_of_week  num_of_guests        hour  table  dessert  not_dessert  Bill  \
id                                                                              
872    Saturday              9        noon     13     True        False     2   
916      Friday              4  after-noon     18    False         True     2   
462      Friday              3     evening     21    False         True     2   
719      Friday              2  after-noon      4    False         True     2   
692      Friday              4       night     20    False         True     2   

     bill  
id         
872   148  
916   169  
462    75  
719    22  
692    27

In [172]:
df_working_days = df[(df.day_of_week != "Friday") & (df.day_of_week != "Saturday")]  # df[~df.day_of_week.isin(["Friday", "Saturday"])]
df_working_days.head()

day_of_week  num_of_guests     hour  table  dessert  not_dessert  Bill  \
id                                                                           
754     Tuesday              4    night      4     True        False     1   
105      Sunday              2  evening      3     True        False     1   
958      Monday             10  evening     17     True        False     2   
49      Tuesday              3  evening      7     True        False     2   
668      Monday              2  evening     22     True        False     2   

     bill  
id         
754    24  
105    60  
958    87  
49    192  
668    28

### @@@ Extras

In [173]:
# Query
df.query("Bill > 150 and hour == 'evening'")
# df[(df.Bill > 50) & (df.hour == 'evening')]

Empty DataFrame
Columns: [day_of_week, num_of_guests, hour, table, dessert, not_dessert, Bill, bill]
Index: []

In [174]:
df.query?

# The _GroupBy()_ class

_pandas_ supports many SQL-like featrues, including Boolean indexing (WHERE/HAVING), join() (JOIN), append() (UNION), sort_values() (ORDER BY) and many others. One of the most important features is the [_GroupBy_ class][1], which is a superset of  the GROUP BY command in SQL. It supports a neat **Split-Apply-Combine** mechanism which facilitates the application of (mostly) aggregative functions on sub-groups of the data.

The _GroupBy_ Split-Apply-Combine mechanism has a fairly simple 2-step call.

* The 1st step is to call _groupby(key)_. This will **split** the data by the specified _key_, and return a _GroupBy_ object.
* The 2nd step is to call some aggregative function on the object. This will **apply** the function to each split and **combine** the results back.

The returned value after the "combine" step is a new _DataFrame_, with the same rows as the original one, except for the _key_ column(s) which become the _index_.

[1]: https://pandas.pydata.org/pandas-docs/stable/reference/groupby.html "The GroupBy class"

In [175]:
df = pd.read_csv("dessert.csv", index_col="id")
df.head()

day.of.week  num.of.guests        hour  table  dessert
id                                                       
1       Monday              2     evening     13     True
2     Saturday              4       night     19     True
3     Saturday              4  after-noon     12    False
4     Thursday              3       night      5     True
5       Monday              7        noon     10     True

In [176]:
pd.read_csv?

In [177]:
df.info

<bound method DataFrame.info of      day.of.week  num.of.guests        hour  table  dessert
id                                                         
1         Monday              2     evening     13     True
2       Saturday              4       night     19     True
3       Saturday              4  after-noon     12    False
4       Thursday              3       night      5     True
5         Monday              7        noon     10     True
...          ...            ...         ...    ...      ...
996       Friday              3  after-noon      5    False
997    Wednesday              5     evening     14     True
998     Thursday              2     evening     20     True
999     Saturday              2     evening      3     True
1000      Friday              2     evening      9    False

[1000 rows x 5 columns]>

In [178]:
grpby_dow = df.groupby('day.of.week')
grpby_dow

In [179]:
print(type(grpby_dow))

<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


In [180]:
list(grpby_dow.groups.keys())

['Friday', 'Monday', 'Saturday', 'Sunday', 'Thursday', 'Tuesday', 'Wednesday']

In [181]:
df['num.of.guests'].sum()

3255

In [182]:
(
     df                       # dataframe      
    .groupby('day.of.week')  # grouping criteria
    ["num.of.guests"]        # which columns
    .sum()                   # what agregation function? # returnes series
)  


day.of.week
Friday       739
Monday       337
Saturday     714
Sunday       301
Thursday     346
Tuesday      401
Wednesday    417
Name: num.of.guests, dtype: int64

It is possible to groub by more than one column:

In [183]:
grpby_dow_sum = df.groupby(['day.of.week', "hour"])["num.of.guests"].sum()
grpby_dow_sum

day.of.week  hour      
Friday       after-noon     77
             evening       338
             night         122
             noon          202
Monday       after-noon     30
             evening       154
             night          53
             noon          100
Saturday     after-noon     56
             evening       276
             night         138
             noon          244
Sunday       after-noon     49
             evening       127
             night          58
             noon           67
Thursday     after-noon     29
             evening       150
             night          56
             noon          111
Tuesday      after-noon     42
             evening       161
             night          86
             noon          112
Wednesday    after-noon     39
             evening       191
             night          65
             noon          122
Name: num.of.guests, dtype: int64

In [184]:
grpby_dow_sum.index

MultiIndex([(   'Friday', 'after-noon'),
            (   'Friday',    'evening'),
            (   'Friday',      'night'),
            (   'Friday',       'noon'),
            (   'Monday', 'after-noon'),
            (   'Monday',    'evening'),
            (   'Monday',      'night'),
            (   'Monday',       'noon'),
            ( 'Saturday', 'after-noon'),
            ( 'Saturday',    'evening'),
            ( 'Saturday',      'night'),
            ( 'Saturday',       'noon'),
            (   'Sunday', 'after-noon'),
            (   'Sunday',    'evening'),
            (   'Sunday',      'night'),
            (   'Sunday',       'noon'),
            ( 'Thursday', 'after-noon'),
            ( 'Thursday',    'evening'),
            ( 'Thursday',      'night'),
            ( 'Thursday',       'noon'),
            (  'Tuesday', 'after-noon'),
            (  'Tuesday',    'evening'),
            (  'Tuesday',      'night'),
            (  'Tuesday',       'noon'),
            ('We

In [185]:
# Extra - accessing using multi-index
grpby_dow_sum[('Wednesday', 'night')]

65

In [186]:
df.groupby('day.of.week').agg(             # agg function
    {
        'num.of.guests': pd.Series.mean,
        'dessert': sum,
        "table": pd.Series.mode,
     }
)

num.of.guests  dessert           table
day.of.week                                        
Friday            3.158120       68               2
Monday            3.209524       70               1
Saturday          3.320930      155               1
Sunday            3.202128       59              19
Thursday          3.392157       73              20
Tuesday           3.208000      110  [1, 8, 10, 13]
Wednesday         3.336000      115              23

The apply-combine steps can be executed on a specific column(s).

In [187]:
grpby_dow.count()

num.of.guests  hour  table  dessert
day.of.week                                     
Friday                 234   234    234      234
Monday                 105   105    105      105
Saturday               215   215    215      215
Sunday                  94    94     94       94
Thursday               102   102    102      102
Tuesday                125   125    125      125
Wednesday              125   125    125      125

In [188]:
grpby_dow.count()['num.of.guests']

day.of.week
Friday       234
Monday       105
Saturday     215
Sunday        94
Thursday     102
Tuesday      125
Wednesday    125
Name: num.of.guests, dtype: int64

In [189]:
grpby_dow['num.of.guests'].count()

day.of.week
Friday       234
Monday       105
Saturday     215
Sunday        94
Thursday     102
Tuesday      125
Wednesday    125
Name: num.of.guests, dtype: int64

It worth noting that the grouping can be also applied using a function. This can be found in the [DataFrame.groupby() constructor](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html#pandas.DataFrame.groupby), and is illustrated below.

In [190]:
def size_grouper(ix):
    size = df.loc[ix, 'num.of.guests']
    if size <= 6:
        return 'Small'
    else:
        return 'Large'

In [191]:
grp_dessert = df.groupby(size_grouper)

In [192]:
grp_dessert.count()

day.of.week  num.of.guests  hour  table  dessert
id                                                     
Large           54             54    54     54       54
Small          946            946   946    946      946

In [193]:
# Even an expression can be used as a grouping criteria
df.groupby(df["num.of.guests"] % 2).count()

day.of.week  num.of.guests  hour  table  dessert
num.of.guests                                                  
0                      567            567   567    567      567
1                      433            433   433    433      433

In [194]:
df.groupby(size_grouper)[['dessert', "num.of.guests"]].sum()

dessert  num.of.guests
id                           
Large       46            443
Small      604           2812

In [195]:
df.groupby(size_grouper)[["num.of.guests"]].sum() # table with one column is series

num.of.guests
id                  
Large            443
Small           2812

In [196]:
# even an expression can be used as grouping criteria
df.groupby(df["num.of.guests"] % 2).count()

day.of.week  num.of.guests  hour  table  dessert
num.of.guests                                                  
0                      567            567   567    567      567
1                      433            433   433    433      433

In [197]:
# This list comprehension generates a list of random integers between 1950 (inclusive) and 2024 (exclusive), 
# with the length equal to the number of rows in the DataFrame.
# for _ in range(len(df))
# The _ is used as a placeholder variable that is not intended to be used in the loop. It's a common convention to use _ for variables that are not needed in the loop body.

df["year"] = [np.random.randint(1950, 2024) for _ in range(len(df))]
df   

day.of.week  num.of.guests        hour  table  dessert  year
id                                                               
1         Monday              2     evening     13     True  2001
2       Saturday              4       night     19     True  1989
3       Saturday              4  after-noon     12    False  2018
4       Thursday              3       night      5     True  1992
5         Monday              7        noon     10     True  2003
...          ...            ...         ...    ...      ...   ...
996       Friday              3  after-noon      5    False  1987
997    Wednesday              5     evening     14     True  2004
998     Thursday              2     evening     20     True  1999
999     Saturday              2     evening      3     True  1965
1000      Friday              2     evening      9    False  1963

[1000 rows x 6 columns]

In [198]:
(df.year // 10) * 10 # generates decade year

id
1       2000
2       1980
3       2010
4       1990
5       2000
        ... 
996     1980
997     2000
998     1990
999     1960
1000    1960
Name: year, Length: 1000, dtype: int64

> **Your turn:**
> * Find the names which are both boy and girl names.
> * How many distinct names are there per year?
> * What is the most common name per year?


In [199]:
df_names = pd.read_csv('baby_names.csv')

In [200]:
df_names.sample(5)

year       name      prop   sex soundex
187675  1938      Robin  0.000091  girl    R150
185438  1936  Annabelle  0.000201  girl    A514
86058   1966      Wayne  0.002690   boy    W500
40987   1920      Ewell  0.000045   boy    E400
105105  1985       Luke  0.001425   boy    L200

### Solution

In [201]:
# Find the names which are both boy and girl names.

set(df_names[df_names.sex == "boy"].name.unique()) & set(df_names[df_names.sex == "girl"].name.unique())

{'Aaron',
 'Abbie',
 'Ada',
 'Addie',
 'Addison',
 'Adrian',
 'Agnes',
 'Akira',
 'Alba',
 'Albert',
 'Alex',
 'Alexander',
 'Alexis',
 'Alfred',
 'Ali',
 'Alice',
 'Allie',
 'Allison',
 'Allyn',
 'Alma',
 'Alpha',
 'Alta',
 'Alva',
 'Amanda',
 'Amari',
 'Amy',
 'Andra',
 'Andrea',
 'Andrew',
 'Angel',
 'Angela',
 'Ann',
 'Anna',
 'Annie',
 'Anthony',
 'Antonia',
 'Archie',
 'Ardell',
 'Arden',
 'Ariel',
 'Arley',
 'Arlie',
 'Armani',
 'Arthur',
 'Artie',
 'Ashley',
 'Ashton',
 'Atha',
 'Aubrey',
 'Audie',
 'Audrey',
 'Augusta',
 'Augustine',
 'Austin',
 'Avery',
 'Baby',
 'Bailey',
 'Barbara',
 'Beatrice',
 'Bee',
 'Bell',
 'Bennie',
 'Bernice',
 'Bert',
 'Bertha',
 'Bertie',
 'Beryl',
 'Bessie',
 'Bethel',
 'Betty',
 'Beulah',
 'Beverly',
 'Billie',
 'Billy',
 'Bird',
 'Blair',
 'Blake',
 'Blanche',
 'Bobbie',
 'Bobby',
 'Bonnie',
 'Brandon',
 'Brandy',
 'Brenda',
 'Brett',
 'Brian',
 'Britt',
 'Brittany',
 'Brook',
 'Burnice',
 'Byrd',
 'Callie',
 'Cameron',
 'Camille',
 'Campbell',

In [202]:
# How many distinct names are there per year?
df_names.groupby("year").name.apply(lambda per_year_names: per_year_names.nunique())

year
1880    1884
1881    1885
1882    1898
1883    1884
1884    1886
        ... 
2004    1931
2005    1935
2006    1941
2007    1938
2008    1936
Name: name, Length: 129, dtype: int64

In [203]:
# What is the most common name per year?
(
    df_names
    .groupby("year")
    .apply(lambda year_group: year_group[year_group.prop == year_group.prop.max()].name)
    .sample(20)
)

year        
1888  8000         John
1936  56000      Robert
1971  91000     Michael
1954  74000     Michael
1880  0            John
1943  63000       James
1890  10000        John
1957  77000     Michael
1908  28000        John
1905  25000        John
1974  94000     Michael
1963  83000     Michael
1992  112000    Michael
1949  198000      Linda
1965  85000     Michael
1940  60000       James
1953  73000      Robert
1904  24000        John
1947  196000      Linda
1899  19000        John
Name: name, dtype: object

## [optional] Additional GroupBy operations

In the apply step, we might wish to do one of the following:
* **Aggregation (`apply()`):** compute a summary statistic (or statistics) for each group. (This is the type of examples we've just seen)
* **Transformation (`transform()`):** perform some group-specific computations and return a like-indexed object.
* **Filtration (`filter()`):** discard some groups, according to a group-wise computation that evaluates True or False.

We will see examples of these functionalities in our pandas practice later in this course.

In [204]:
df = pd.read_csv("dessert.csv", index_col="id")
df.head()

day.of.week  num.of.guests        hour  table  dessert
id                                                       
1       Monday              2     evening     13     True
2     Saturday              4       night     19     True
3     Saturday              4  after-noon     12    False
4     Thursday              3       night      5     True
5       Monday              7        noon     10     True

In [205]:
# .apply
# Argument to .apply is a function which takes a set of rows, a group.
# In statistics "mode" refers to a value which appears most often.
def mode(df_single_group):
  return df_single_group["num.of.guests"].value_counts().index[0]

df.groupby("day.of.week").apply(mode)

day.of.week
Friday       2
Monday       2
Saturday     2
Sunday       2
Thursday     2
Tuesday      2
Wednesday    2
dtype: int64

In [206]:
# .transform
# Similar to .apply, operating per column. Index structure must be maintained.
# Take care to select the columns your function expects to work on.
# In statistics, "ptp" (peak-to-peak) means the range of values in a given set of numbers (i.e., max - min).
df["dow.ptp"] = df.groupby("day.of.week")["num.of.guests"].transform(lambda column_series: column_series.max() - column_series.min())
df.sample(10)

day.of.week  num.of.guests        hour  table  dessert  dow.ptp
id                                                                 
366    Thursday              4     evening     16    False        9
325      Sunday              3        noon     22     True        9
432      Sunday              2       night      6     True        9
874      Monday              4  after-noon     20    False        9
866      Friday              3  after-noon     20    False        9
175    Saturday              3        noon     11     True        8
452      Sunday              4        noon      2     True        9
811      Friday              3     evening      6    False        9
514      Friday             10       night      3     True        9
353      Friday              2       night     19     True        9

In [207]:
# .filter # HAVING IN SQL
# Filter groups based on some criteria, as calculated by a function (similar to HAVING clause in SQL)
# df.groupby("day.of.week")["num.of.guests"].max()
df.groupby("day.of.week").filter(lambda df_single_group: df_single_group["num.of.guests"].max() > 9)

day.of.week  num.of.guests        hour  table  dessert  dow.ptp
id                                                                  
1         Monday              2     evening     13     True        9
4       Thursday              3       night      5     True        9
5         Monday              7        noon     10     True        9
6         Friday              2     evening     19    False        9
7         Monday              2     evening     22     True        9
...          ...            ...         ...    ...      ...      ...
995       Friday              1        noon      8    False        9
996       Friday              3  after-noon      5    False        9
997    Wednesday              5     evening     14     True        9
998     Thursday              2     evening     20     True        9
1000      Friday              2     evening      9    False        9

[785 rows x 6 columns]

# Exercise <--- Homework 25/7/2024

The dataset `diamonds.csv` contains data on diamonds, their attributes and their prices. Familiarize yourself with the dataset (you can have a quick look here if you'd like to understand the meaning behind some of the column names https://www.gia.edu/diamond-quality-factor) and answer the following questions:

1. What is the most expensive diamond?
2. What is the carat of the most Expensive diamond?
3. What is the mean price of each diamond color?
4. What is the max carat for each diamon cut?
5. What is the most expensive combination of depth and clarity?
6. What is the most common carat?
7. What is the mean carat for clarity SI2 and depth greater than 60?

In [208]:
dataset = pd.read_csv('diamonds.csv')
dataset.head()

#  carat      cut color clarity  depth  table  price     x     y     z
0  1   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1  2   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2  3   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3  4   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4  5   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

In [209]:
# 1. What is the most expensive diamond?
dataset.sort_values(by = 'price', ascending=False).head(1)


#  carat      cut color clarity  depth  table  price    x     y  \
27749  27750   2.29  Premium     I     VS2   60.8   60.0  18823  8.5  8.47   

          z  
27749  5.16

In [210]:
# 2. What is the carat of the most Expensive diamond?
dataset.sort_values(by = 'price', ascending=False).head(1)['carat']

27749    2.29
Name: carat, dtype: float64

In [211]:
# 3. What is the mean price of each diamond color?
dataset.groupby('color')['price'].mean().round(2)

color
D    3169.95
E    3076.75
F    3724.89
G    3999.14
H    4486.67
I    5091.87
J    5323.82
Name: price, dtype: float64

In [213]:
# 4. What is the max carat for each diamon cut?
dataset.groupby('cut')['carat'].max()

cut
Fair         5.01
Good         3.01
Ideal        3.50
Premium      4.01
Very Good    4.00
Name: carat, dtype: float64

In [212]:
# 5. What is the most expensive combination of depth and clarity?
dataset.groupby(['depth', 'clarity'])['price'].max().sort_values(ascending=False).head(1)

depth  clarity
60.8   VS2        18823
Name: price, dtype: int64

In [214]:
# 6. What is the most common carat?
dataset.groupby('carat').x.count().sort_values().tail(1).index[0]

# Another option:
dataset.groupby('carat').size().sort_values().tail(1).index[0]

0.3

In [197]:
# 7. What is the mean carat for clarity SI2 and depth greater than 60?
dataset.loc[(dataset['clarity']=='SI2') & (dataset['depth']>60)].carat.mean()
         


1.0724980988593247

## Solution

In [ ]:
# 1. What is the most expensive diamond?
# 2. What is the carat of the most Expensive diamond?
dataset.sort_values(by='price', ascending=False).head(1)

In [ ]:
# 1. What is the most expensive diamond? option 2
dataset['price'].idxmax()

In [ ]:
dataset.loc[27749]

In [ ]:
# 3. What is the mean price of each diamon color?
dataset.groupby('color')['price'].mean()

In [ ]:
# 4. What is the max carat for each diamon cut?
dataset.groupby('cut')['carat'].max()

In [ ]:
# 5. What is the most expensive combination of depth and clarity?
dataset.groupby(['depth','clarity'])['price'].max().sort_values(ascending=False).head(1)

In [ ]:
# 6. What is the most common carat?
dataset.groupby('carat').x.count().sort_values().tail(1).index[0]

# Another option:
dataset.groupby('carat').size().sort_values().tail(1).index[0]

In [196]:
# 7. What is the mean carat for clarity SI2 and depth greater than 60?
dataset[(dataset['clarity'] == 'SI2') & (dataset['depth'] > 60)].carat.mean()

1.0724980988593247

# Example - US baby names (more)

The file "Baby names.csv" contains information about the most frequent names in the US. For each year from 1880 to 2008 it lists the 1000 most frequent names for both boys and girls (129 x 2 x 1000 = 258000 rows). Each row includes the year, the name, the sex, the [soundex][1] of the name (which we will ignore), and most importantly - the name's frequency among its gender. We will use _pandas_ to explore this data.

[1]: https://en.wikipedia.org/wiki/Soundex "soundex - Wikipedia"

> **NOTE:** This example is fully based on a classic presentation given by _pandas_ main author Wes McKinney at 2012. The full presentation can be found on [YouTube][1]. The transcription itself is taken from [Jennifer Lee's GitHub page][2].

[1]: https://www.youtube.com/watch?v=w26x-z-BdWQ "Wes McKinney - Introdcution to pandas"
[2]: https://github.com/estimate/pandas-exercises "Jennifer Lee's GitHub page"

## Exploration

In [ ]:
df_names = pd.read_csv('baby_names.csv', nrows=500000)

In [ ]:
df_names.head(5)

In [ ]:
df_names.tail(3)

In [ ]:
df_names[df_names.year == 1880].tail()

We separate the data into two data frames - _boys_ and _girls_. We will do the exploration with the boys, and then apply the relevant scripts to the girls as well.

In [ ]:
df_boys = df_names[df_names.sex == 'boy']
df_girls = df_names[df_names.sex == 'girl']

The sum of the 'prop' column for each year does not sum to 1.

In [ ]:
df_boys.head()

In [ ]:
boys_grouped = df_boys.groupby('year')
boys_grouped

In [ ]:
(
  df_boys
  .groupby('year')
  ['prop']
  .sum()
  .plot()
)

What does the graph demonstrate?

## Part I

**Task:** We wish to visualize the changes in the popularity of the most popular name through the years.

Our approach will be to group the data by year, then to extract from each group the record of the most frequent name, and then to plot the combined result.

In [ ]:
(
    df_boys
    .groupby('year')
    ['prop']
    .max()
    .plot()
)

In [ ]:
most_popular_indexes = (
    df_boys
    .groupby("year")
    .prop
    .idxmax()
)

df_boys.loc[most_popular_indexes]

Can you see the trend?

And finally...

In [ ]:
# Same as above, but a slightly different approach

def get_max_record(group):
    return group.prop.max()

In [ ]:
most_popular = df_boys.groupby('year').apply(get_max_record)
most_popular.head()

In [ ]:
most_popular.plot(title='Highest boys prop through years')

In [ ]:
most_popular = df_girls.groupby('year').apply(get_max_record)
most_popular.plot(title='Highest girls prop through years')

## Part II

**Task:** Visualize the trend of a specific name through the years.

In [ ]:
name = 'David'
df_boys.loc[df_boys.name == name].set_index('year').prop.plot()
# same as: df_boys.loc[df_boys.name==name].set_index('year')["prop"].plot()
# Column name, prop, could be omitted becuase after set_index, the only other numeric column is prop

In [ ]:
# same graph with explicit setting of X and Y
df_boys[df_boys.name==name].plot(x="year", y="prop")

We can group by the name, but since we are not interested in the data of the other names, this is a wasteful approach. We can either slice the data of the relevant name, or reindex by the name. Also, since plots use the index as their x axis, it will be simpler to make the year a secondary index.

In [ ]:
name = 'Elvis'

In [ ]:
df_boys2 = df_boys.set_index(['name', 'year'])
# print(df_boys2.head())
df_boys2.loc[name].prop.plot()

## Part III

**Task:** Visualize the "diversity" - how many names are required in order to describe half of the population - through the years.

For each year we will sum the 'prop' column (using the _cumsum()_ function) and find where this sum exceeds 0.5 (using the function [_searchsorted()_][1]).

[1]: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.searchsorted.html

### Demonstration

In [ ]:
df_1880 = df_boys[df_boys.year == 1880]
df_1880.prop.head()

In [ ]:
df_1880.prop.cumsum().head()

In [ ]:
df_1880.prop.cumsum().searchsorted(0.5)

In [ ]:
# Generalize for all the years and plot
(
  df_girls
  .groupby('year')
  ['prop']
  .apply(lambda year_prop_ser: year_prop_ser.cumsum().searchsorted(0.5))
  .plot()
)

### Solution

In [ ]:
def get_diversity(group):
    return group.prop.cumsum().searchsorted(0.5)

In [ ]:
df_boys.groupby('year').apply(get_diversity).plot()

In [ ]:
ax = df_boys.groupby('year').apply(get_diversity).plot(linewidth=3)
ax = df_girls.groupby('year').apply(get_diversity).plot(linewidth=3, ax=ax)
ax.legend(['boys', 'girls'], loc='best')

In [ ]:
# Another option
df_names.groupby(["year", "sex"]).apply(get_diversity).unstack(1).plot(linewidth=3)

# Visualization

> **Note:** matplotlib is the visualization engine of pandas, and although most of our work will be done directly with pandas, it is a good idea to get acquainted with matplotlib basics through the [chapter about matplotlib](https://drive.google.com/drive/folders/1lwVdMjZva7OpPPYFDEocmvic8bib_x2K?usp=sharing). More specifically, one should be familiar with the concepts of Figure, Axes, the `plot` method and the varios advanced types of plot, like histogram, bars, scatter, etc.

The DataFrame class supports the `plot` accessor, which is a simple wrapper for many of the matplotlib plotting functions.
We will illustrate some of them here, using [Kaggle's bike sharing dataset](https://www.kaggle.com/c/bike-sharing-demand).

We now read the data. The first column is a datetime column, so we indicate that using the [`parse_dates`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) argument.

<pre>
Data Fields
datetime - hourly date + timestamp  
season -  1 = spring, 2 = summer, 3 = fall, 4 = winter
holiday - whether the day is considered a holiday
workingday - whether the day is neither a weekend nor holiday
weather - 1: Clear, Few clouds, Partly cloudy, Partly cloudy
2: Mist + Cloudy, Mist + Broken clouds, Mist + Few clouds, Mist
3: Light Snow, Light Rain + Thunderstorm + Scattered clouds, Light Rain + Scattered clouds
4: Heavy Rain + Ice Pallets + Thunderstorm + Mist, Snow + Fog
temp - temperature in Celsius
atemp - "feels like" temperature in Celsius
humidity - relative humidity
windspeed - wind speed
casual - number of non-registered user rentals initiated
registered - number of registered user rentals initiated
count - number of total rentals
</pre>

In [ ]:
bikes = pd.read_csv('bike sharing train.csv', parse_dates=[0])
bikes.sample(10)

In [ ]:
bikes["datetime"] = bikes["datetime"].astype("datetime64")

In [ ]:
bikes.info()

In [ ]:
bikes.describe()

> **Note:** The graphs below serve a short demonstration. We will have a thorough EDA of this dataset later in the course.

The `hist()` method is a great visualization to stat with.

In [ ]:
ax = bikes.hist(figsize=(12, 12))

**Example 1** - Throughout the day and night hours, what is the difference between the usage of registered users and casual users? Is there a change considering the working day?

In [ ]:
bikes.datetime.max() - bikes.datetime.min()

In [ ]:
s_hour = bikes['datetime'].dt.hour
s_hour

In [ ]:
bikes.groupby(s_hour)[['casual', 'registered']].mean().head(3)

In [ ]:
# Overall difference in usage between registered and casual users (i.e., no division between workdays and weekends)
bikes.groupby(s_hour)[['casual', 'registered']].mean().plot(figsize=(9, 6), lw=3)

In [ ]:
# Same as above, but this time dividing between workdays and weekends
bikes.groupby(['workingday', s_hour])[['casual', 'registered']].mean().plot(figsize=(9, 6), lw=3)

In [ ]:
# Advanced - multi-index manipulation
(
  bikes
 .groupby(['workingday', s_hour])
 [['casual', 'registered']]
 .mean()
 .unstack(0)
 .plot(figsize=(9, 6), lw=3)
)

In [ ]:
# Just for the fun of it, lets split the four plots into a grid of individual plots
bikes.groupby(['workingday', s_hour])[['casual', 'registered']].mean().unstack(0)\
  .plot(
      subplots=True,              # each column gets it's own plot
      layout=(2, 2),              # subplots grid layout
      sharex=True, sharey=True,   # same X and Y axes for all the plots
      figsize=(10, 7)
  )

**Example 2** - What is the influence of workingday on the distribution of cyclists (i.e., distribution of the number of cyclists)?

In [ ]:
ax = bikes.hist(column='casual', by='workingday', bins=50, sharex=True, sharey=True, figsize=(10, 4))

In [ ]:
ax = bikes.hist(column='registered', by='workingday', bins=50, sharex=True, sharey=True, figsize=(10, 4))

In [ ]:
print("casual % on working days:", bikes[bikes.workingday == 1].casual.sum() / bikes.casual.sum())
print("registered % on working days:", bikes[bikes.workingday == 1].registered.sum() / bikes.registered.sum())

> **Note:** Example 2 can be tried out with `weather` and `season`

**Example 3** - Given an hour of the day, what is the relation between the temperature and the number of cyclists?

In [ ]:
bikes.datetime.dt.hour

In [ ]:
hour = 9
bikes[bikes['datetime'].dt.hour == hour].plot\
  .scatter(
      x='temp', y='count',
      c='red',
      figsize=(9, 6)
)

In [ ]:
# BTW, if you want to highlight densities (i.e., areas in the scatter where more dots are clustered),
# "hexbin" plot may be useful.
hour = 9
(
  bikes
  [bikes['datetime'].dt.hour==hour]
  .plot
  .hexbin(
      x='temp', y='count',
      gridsize=15,
      cmap="Reds",
      figsize=(9, 6)
  )
)

In [ ]:
# splitting between "casual" and "registered" users
hour = 9
ax = bikes[bikes['datetime'].dt.hour == hour].plot.scatter(x='temp', y='registered', c='red')
bikes[bikes['datetime'].dt.hour == hour].plot.scatter(x='temp', y='casual', c='blue', ax=ax)
leg = ax.legend(['registered', 'casual'])

> **Your turn:** Plot the following for the dessert dataset:
> * Plot 1 - A bar-plot indicating the number of buyers and non-buyers for each group size
> * Plot 2 - A pie chart indicating the relative part of each weekday in the data

#### Solution

In [ ]:
df = pd.read_csv("dessert.csv")
df.head()

In [ ]:
# Plot 1 - A bar-plot indicating the number of buyers and non-buyers for each group size
df["no.dessert"] = ~df["dessert"]
df.groupby("num.of.guests")[["dessert", "no.dessert"]].sum()

In [ ]:
df.groupby("num.of.guests")[["dessert", "no.dessert"]].sum().plot.bar(figsize=(9, 6))   # try it using barh (just like in Matplotlib)

In [ ]:
# Eddy's solution
df.groupby(['num.of.guests','dessert'])['id'].count().unstack(1).plot.bar()

In [ ]:
# Plot 2 - A pie chart indicating the relative part of each weekday in the data
df['day.of.week'].value_counts()

In [ ]:
df['day.of.week'].value_counts().plot.pie(figsize=(10, 7), autopct="%.1f")

### Worth mentioning...

In [ ]:
df.pivot_table?

In [ ]:
df.melt?